apps/src/templates/projects/projectsRedux.js

```diff
@@ -302,6 +322,57 @@ export function updateProjectName(projectId, updatedName) {
   return {type: UPDATE_PROJECT_NAME, projectId, updatedName};
 }
 
+export function saveSuccess(projectId) {
+  return {type: SAVE_SUCCESS, projectId};
+}
+
+export function saveFailure(projectId) {
+  return {type: SAVE_FAILURE, projectId};
+}
+
+const fetchProjectToUpdate = (projectId, onComplete) => {
+  $.ajax({
+    url: `/v3/channels/${projectId}`,
+    method: 'GET',
+    type: 'json',
+    contentType: 'application/json;charset=UTF-8',
+  }).done((data) => {
+    onComplete(null, data);
+  }).fail((jqXhr, status) => {
+    onComplete(status, jqXhr.responseJSON);
+  });
+};
+
+const updateProjectNameOnServer = (project) => {
+  return (dispatch) => {
+    $.ajax({
+      url: `/v3/channels/${project.id}`,
+      method: 'POST',
+      type: 'json',
+      contentType: 'application/json;charset=UTF-8',
+      data: JSON.stringify(project)
+    }).done((data) => {
+      dispatch(saveSuccess(project.id));
+    }).fail((jqXhr, status) => {
+      dispatch(saveFailure(project.id));
+    });
+  };
+};
+
+export const saveProjectName = (projectId, updatedName) => {
+  return (dispatch) => {
+    fetchProjectToUpdate(projectId,
```

Comment:
Why do we need to fetch the project before we can update it?

Reply:
It's my understanding, based on `channels_api`, that updating the project requires all of the project data, not just the field we want to update. Also personal projects have a slightly different shape than a project returned directly from `/v3/channels/<channel-id>`. 
<img width="1096" alt="screen shot 2018-08-06 at 4 05 27 pm" src="https://user-images.githubusercontent.com/12300669/43745366-2d331d8e-9993-11e8-9fc6-cc55ee2ac22f.png">
Left is a personal project, right is what `/v3/channels/<channel-id>` returns.
So, if I want to accurately update the project, I need to fetch it and pass the expected information when I make the call to update. Am I confused?

Reply:
@islemaster that's a good point, i hadn't considered the possibility of an out-of-order write. your suggestion is definitely safer, and cuts down on API calls, which is nice.

i wonder which route would be better: 1) update the existing endpoint to accept just an id and the new project name, or 2) add a specific route that is only for updating a project name (sort of like the [set_user_type](https://github.com/code-dot-org/code-dot-org/blob/c6097e9aeec61e05ce2c357386346e1b9de52fb0/dashboard/app/controllers/registrations_controller.rb#L193-L226) and [set_email](https://github.com/code-dot-org/code-dot-org/blob/c6097e9aeec61e05ce2c357386346e1b9de52fb0/dashboard/app/controllers/registrations_controller.rb#L150-L185) endpoints in registrations_controller).

Reply:
I think it depends on whether we've got a good sense of every place the existing API is used.  It seems like the preferable outcome would be (1) where the API behaves more like a Rails model update, only mutating fields you provide; but (2) is the safer choice, guaranteed not to break any existing consumers.

We might be able to get the best of both worlds by introducing a `PATCH /v3/channels/<project>` route with the behavior of (1) and gradually migrating over to it.

I also want to acknowledge that this might make the most sense in a follow-up PR, and the existing solution is probably good enough to roll out the React version of the project list. :tada:

Reply:
I'm leaning toward adding a specific route for renaming because I don't want to break all of the other places where the existing endpoint for updating a project is uses by modifying it too significantly.